internal/provider/providers/namesilo/provider.go

```diff
@@ -0,0 +1,305 @@
+		Owner:     p.Owner(),
+		Provider:  "<a href=\"https://www.namesilo.com/\">NameSilo</a>",
+		IPVersion: p.ipVersion.String(),
+	}
+}
+
+// Update does the following:
+// 1. if there's no record, create it.
+// 2. if it exists and ip is different, update it.
+// 3. if it exists and ip is the same, do nothing.
+func (p *Provider) Update(ctx context.Context, client *http.Client, newIP netip.Addr) (netip.Addr, error) {
+	recordType := constants.A
+	if newIP.Is6() {
+		recordType = constants.AAAA
+	}
+
+	// retrieve the current DNS record for the given IP type
+	recordID, currentIP, err := p.getRecord(ctx, client, recordType)
+	if err != nil && !stderrors.Is(err, errors.ErrRecordNotFound) {
+		return netip.Addr{}, fmt.Errorf("error retrieving records for %s: %w", p.domain, err)
+	}
+
+	// if the current IP is different from the new IP, update the record
+	if currentIP != newIP {
+		// pass the recordID for updating, or nil for creating a new record
+		if err := p.createOrUpdateRecord(ctx, client, recordID, recordType, newIP); err != nil {
+			return netip.Addr{}, fmt.Errorf("error setting record for %s: %w", p.BuildDomainName(), err)
+		}
+	}
+
+	// return the new IP (whether updated or unchanged)
+	return newIP, nil
+}
+
+// https://www.namesilo.com/api-reference#dns/dns-list-records
+func (p *Provider) getRecord(ctx context.Context, client *http.Client, recordType string) (
+	id *string, ip netip.Addr, err error,
+) {
+	queryParams := url.Values{}
+	url := p.createRequestURL("/api/dnsListRecords", queryParams)
+
+	response, err := p.sendAPIRequest(ctx, client, url)
+	if err != nil {
+		return nil, netip.Addr{}, err
+	}
+
+	// find matching record
+	host := utils.BuildURLQueryHostname(p.owner, p.domain)
+	for _, record := range response.Reply.Records {
+		if record.Host != host || record.Type != recordType {
+			continue
+		}
+		ip, err = netip.ParseAddr(record.Value)
+		if err != nil {
+			return nil, netip.Addr{}, fmt.Errorf("parsing existing IP: %w", err)
+		}
+		return &record.ID, ip, nil
+	}
+
+	return nil, netip.Addr{}, fmt.Errorf("%w: no matching records found", errors.ErrRecordNotFound)
+}
+
+// https://www.namesilo.com/api-reference#dns/dns-add-record
+// https://www.namesilo.com/api-reference#dns/dns-update-record
+func (p *Provider) createOrUpdateRecord(
+	ctx context.Context,
+	client *http.Client,
+	recordID *string,
+	recordType string,
+	ip netip.Addr,
+) error {
+	name := p.owner
+	if name == "@" {
+		name = ""
+	}
+
+	queryParams := url.Values{
+		"rrhost":  {name},
+		"rrvalue": {ip.String()},
+	}
+	if p.ttl != nil {
+		queryParams.Set("rrttl", strconv.FormatUint(uint64(*p.ttl), 10))
+	}
+
+	var path string
+	if recordID == nil {
+		// create new record
+		path = "/api/dnsAddRecord"
+		queryParams.Set("rrtype", recordType)
+	} else {
+		// update record by id
+		path = "/api/dnsUpdateRecord"
+		queryParams.Set("rrid", *recordID)
+	}
```

Comment:
Let's split this in two functions, one `createRecord` and one `updateRecord`. So the top level Update function can just act depending if the record not found error is encountered, it would feel more structured I think.